File: src/front-end/fun-chess-app/src/core/chess/pieces/King.ts
import Piece from "../Piece";
import Board from "../Board";
import Cell from "../Cell";
import CastlingPlay from "../enums/CastlingPlay";
import SpecialMove from "../enums/SpecialMove";
import Team from "../enums/Team";
import Move from "../structs/Move";
import Position from "../structs/Position";
import { checkEast, checkWest } from "../utils/cells-utils";
import Rook from "./Rook";

const MOVE_LENGTH = 1;

class King extends Piece {
    private constructor() {
        super(4);
    }

    public static readonly instance: Piece = new King();

    public override move(board: Board, move: Move): [boolean, SpecialMove] {
        const tuple = super.move(board, move);
        if (!tuple[0]) return tuple;
        
        board.teams.get(board.turn)!.castlingPlays = CastlingPlay.None;
        board.teams.get(board.turn)!.kingPosition = move.next;
        
        switch (tuple[1]) {
            case SpecialMove.LeftCastling:
                switch (board.turn) {
                    case Team.White:
                        board.internalBoard[3] = board.internalBoard[0];
                        board.internalBoard[0] = Cell.Empty;
                        break;
                    case Team.Black:
                        board.internalBoard[59] = board.internalBoard[56];
                        board.internalBoard[56] = Cell.Empty;
                        break;
                }
                return tuple;
            case SpecialMove.RightCastling:
                switch (board.turn) {
                    case Team.White:
                        board.internalBoard[5] = board.internalBoard[7];
                        board.internalBoard[7] = Cell.Empty;
                        break;
                    case Team.Black:
                        board.internalBoard[61] = board.internalBoard[63];
                        board.internalBoard[63] = Cell.Empty;
                        break;
                }
                return tuple;
            default:
                return tuple;
        }
    }

    public override moveIsValid(board: Board, move: Move): [boolean, SpecialMove] {
        let specialMove = SpecialMove.None;
        const moveIsValid = move.diffX >= -MOVE_LENGTH && move.diffX <= MOVE_LENGTH && move.diffY >= -MOVE_LENGTH && move.diffY <= MOVE_LENGTH;
        const possiblePlays = board.teams.get(board.turn)!.castlingPlays;
        
        if (moveIsValid || possiblePlays === CastlingPlay.None || move.diffY !== 0 || board.kingInDanger()) return [moveIsValid, specialMove];
        
        let previous: Position, next: Position;
        switch (move.diffX) {
            case 2:
                specialMove = SpecialMove.RightCastling;
                if ((possiblePlays & CastlingPlay.RightCastling) === CastlingPlay.None) return [false, specialMove];
                if (!checkEast(board.internalBoard, move)) return [false, specialMove];

                previous = new Position(board.turn === Team.White ? 7 : 63);
                next = new Position(board.turn === Team.White ? 5 : 61);
                return [Rook.instance.moveIsValid(board, new Move(previous, next))[0], specialMove];
            case -2:
                specialMove = SpecialMove.LeftCastling;
                if ((possiblePlays & CastlingPlay.LeftCastling) === CastlingPlay.None) return [false, specialMove];
                if (!checkWest(board.internalBoard, move)) return [false, specialMove];

                previous = new Position(board.turn === Team.White ? 0 : 56);
                next = new Position(board.turn === Team.White ? 3 : 59);
                return [Rook.instance.moveIsValid(board, new Move(previous, next))[0], specialMove];
        }
        return [false, specialMove];
    }
}

export default King;

File: src/back-end/FunChess/src/FunChess.Core.Chess/Pieces/King.cs
using FunChess.Core.Chess.Enums;
using FunChess.Core.Chess.Extensions;
using FunChess.Core.Chess.Structs;

namespace FunChess.Core.Chess.Pieces;

internal sealed class King : Piece
{
    private const int MoveLength = 1;

    public override bool Move(Board board, Move move)
    {
        if (!MoveIsValid(board, move, out SpecialMove specialMove)) return false;
        
        board.InternalBoard[move.Next.Index] = board.InternalBoard[move.Previous.Index];
        board.InternalBoard[move.Previous.Index] = Cell.Empty;
        board.Teams[board.Turn].CastlingPlays = CastlingPlay.None;
        board.Teams[board.Turn].KingPosition = move.Next;
        
        switch (specialMove)
        {
            case SpecialMove.LeftCastling:
                switch (board.Turn)
                {
                    case Team.White:
                        board.InternalBoard[3] = board.InternalBoard[0];
                        board.InternalBoard[0] = Cell.Empty;
                        break;
                    case Team.Black:
                        board.InternalBoard[59] = board.InternalBoard[56];
                        board.InternalBoard[56] = Cell.Empty;
                        break;
                }
                return true;
            case SpecialMove.RightCastling:
                switch (board.Turn)
                {
                    case Team.White:
                        board.InternalBoard[5] = board.InternalBoard[7];
                        board.InternalBoard[7] = Cell.Empty;
                        break;
                    case Team.Black:
                        board.InternalBoard[61] = board.InternalBoard[63];
                        board.InternalBoard[63] = Cell.Empty;
                        break;
                }
                return true;
            default:
                return true;
        }
    }

    public override bool MoveIsValid(Board board, Move move, out SpecialMove specialMove)
    {
        specialMove = SpecialMove.None;
        bool moveIsValid = move.DiffX is >= -MoveLength and <= MoveLength && move.DiffY is >= -MoveLength and <= MoveLength;
        CastlingPlay possiblePlays = board.Teams[board.Turn].CastlingPlays;
        
        if (moveIsValid || possiblePlays == CastlingPlay.None || move.DiffY != 0 || board.KingInCheck()) return moveIsValid;
        
        Position previous, next;
        switch (move.DiffX)
        {
            case 2:
                specialMove = SpecialMove.RightCastling;
                if ((possiblePlays & CastlingPlay.RightCastling) == CastlingPlay.None) return false;
                if (!board.InternalBoard.CheckEast(move)) return false;

                previous = new Position(board.Turn == Team.White ? (byte)7 : (byte)63);
                next = new Position(board.Turn == Team.White ? (byte)5 : (byte)61);
                return Rook.MoveIsValid(board, new Move(previous, next), out _);
            case -2:
                specialMove = SpecialMove.LeftCastling;
                if ((possiblePlays & CastlingPlay.LeftCastling) == CastlingPlay.None) return false;
                if (!board.InternalBoard.CheckWest(move)) return false;

                previous = new Position(board.Turn == Team.White ? (byte)0 : (byte)56);
                next = new Position(board.Turn == Team.White ? (byte)3 : (byte)59);
                return Rook.MoveIsValid(board, new Move(previous, next), out _);
        }
        return false;
    }
}